src/S3/S3Transfer/S3TransferManager.php

```diff
@@ -0,0 +1,1048 @@
+                $objectsUploaded++;
+
+                return $response;
+            })->otherwise(function ($reason) use (
+                $bucketTo,
+                $sourceDirectory,
+                $failurePolicyCallback,
+                $uploadRequestArgs,
+                $uploadDirectoryRequestArgs,
+                &$objectsUploaded,
+                &$objectsFailed
+            ) {
+                $objectsFailed++;
+                if($failurePolicyCallback !== null) {
+                    call_user_func(
+                        $failurePolicyCallback,
+                        $uploadRequestArgs,
+                        [
+                            ...$uploadDirectoryRequestArgs,
+                            "source_directory" => $sourceDirectory,
+                            "bucket_to" => $bucketTo,
+                        ],
+                        $reason,
+                        new UploadDirectoryResponse(
+                            $objectsUploaded,
+                            $objectsFailed
+                        )
+                    );
+
+                    return;
+                }
+
+                throw $reason;
+            });
+        }
+
+        return Each::ofLimitAll($promises, $this->config['concurrency'])
+            ->then(function ($_) use ($objectsUploaded, $objectsFailed) {
+                return new UploadDirectoryResponse($objectsUploaded, $objectsFailed);
+            });
+    }
+
+    /**
+     * @param string|array $source The object to be downloaded from S3.
+     * It can be either a string with a S3 URI or an array with a Bucket and Key
+     * properties set.
+     * @param array $downloadRequestArgs The getObject request arguments to be provided as part
+     * of each get object operation, except for the bucket and key, which
+     * are already provided as the source.
+     * @param array $config The configuration to be used for this operation:
+     *  - multipart_download_type: (string, optional)
+     *    Overrides the resolved value from the transfer manager config.
+     *  - checksum_validation_enabled: (bool, optional) Overrides the resolved
+     *    value from transfer manager config for whether checksum validation
+     *    should be done. This option will be considered just if ChecksumMode
+     *    is not present in the request args.
+     *  - track_progress: (bool) Overrides the config option set in the transfer
+     *    manager instantiation to decide whether transfer progress should be
+     *    tracked.
+     *  - minimum_part_size: (int) The minimum part size in bytes to be used
+     *    in a range multipart download. If this parameter is not provided
+     *    then it fallbacks to the transfer manager `target_part_size_bytes`
+     *    config value.
+     * @param TransferListener[]|null $listeners
+     * @param TransferListener|null $progressTracker
+     *
+     * @return PromiseInterface
+     */
+    public function download(
```

Comment:
I believe we should have recursive download functionality as well

Reply:
Sorry, what do you mean with recursive download? this API method is intended for single file downloads.